specs/NestedFrame.md

```diff
@@ -0,0 +1,215 @@
+        args.Frame.FrameCreated += (frameCreatedSender, frameCreatedArgs) => 
+        {
+            CoreWebView2Frame childFrame = frameCreatedArgs.Frame;
+            childFrame.NavigationStarting += HandleChildFrameNavigationStarting;
+        }
+    }
+}
+
+void HandleChildFrameNavigationStarting(object sender, 
+    CoreWebView2NavigationStartingEventArgs args)
+{
+    // Manage the navigation, e.g. cancel the navigation 
+    // if it's on block list.
+    CoreWebView2Frame frame = (CoreWebView2Frame)sender;
+    if (!_frameNavigationUrls.ContainsKey(frame.FrameId))
+    {
+        _frameNavigationUrls[frame.FrameId] = new List<string>();
+    }
+    // Log the navigation history per frame Id. 
+    _frameNavigationUrls[frame.FrameId].Add(args.Uri);
+}
+```
+
+# API Details
+## C++
+```C++
+/// Receives `FrameCreated` events.
+interface ICoreWebView2FrameChildFrameCreatedEventHandler : IUnknown {
+  /// Provides the event args for the corresponding event.
+  HRESULT Invoke(
+      [in] ICoreWebView2Frame* sender,
+      [in] ICoreWebView2FrameCreatedEventArgs* args);
+}
+
+/// This is the ICoreWebView2Frame interface.
+interface ICoreWebView2Frame7 : IUnknown {
+  /// Adds an event handler for the `FrameCreated` event.
+  /// Raised when a new direct descendant iframe is created.
+  /// Handle this event to get access to `ICoreWebView2Frame` objects.
+  /// Use `ICoreWebView2Frame::add_Destroyed` to listen for when this
+  /// iframe goes away.
+  /// 
+  /// \snippet ScenarioWebViewEventMonitor.cpp AddFrameCreated
+  HRESULT add_FrameCreated(
+      [in] ICoreWebView2FrameChildFrameCreatedEventHandler* eventHandler,
+      [out] EventRegistrationToken* token);
+
+  /// Removes an event handler previously added with `add_FrameCreated`.
+  HRESULT remove_FrameCreated(
+      [in] EventRegistrationToken token);
+}
+```
+
+## C#
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    runtimeclass CoreWebView2Frame
+    {
+        [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2Frame7")]
+        {
+            event Windows.Foundation.TypedEventHandler<CoreWebView2Frame, CoreWebView2FrameCreatedEventArgs> FrameCreated;
+        }
+    }
+}
+```
+
+# Appendix
+## Impacted API
+### `CoreWebView2Frame.PermissionRequested` and `CoreWebView2Frame.ScreenCaptureStarting`
+In the current case of nested iframes, the [PermissionRequested](https://learn.microsoft.com/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2frame#permissionrequested)
+and [ScreenCaptureStarting](https://learn.microsoft.com/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2frame#screencapturestarting)
+events will be raised from the top-level iframe. With the support
+of tracking nested iframes, we can now handle these requests directly
+within the nested iframe. Specifically, these requests are raised to
+the nearest tracked frame, which is the `CoreWebView2Frame` closest

```

Comment:
This could lead to a problem if an app combines multiple components, and one of them adds a FrameCreated handler and accidentally messes up the other component which expected the event from the top-level frame. Is the answer "Yeah, you can't really componentize your app like that because features used by one app might have adverse consequences far away." So it means I can't have a "app usage telemetry" component that hooks all frame activity and logs it, because that would accidentally break my "permission manager" component.

[From the Windows Runtime design guidelines](https://www.osgwiki.com/wiki/WinRT_API_Design_Guidelines#A_Well-Designed_API_Avoids_Shared_Behavior):

> ### A Well-Designed API Avoids Shared Behavior
> A developer is often unpleasantly surprised when he or she uses one API that causes a side-effect elsewhere.
> 
> ☑ DO design APIs to allow multiple components running in the same process to use the API without having to explicitly coordinate, for example, two plug-ins loaded into the same instance of Internet Explorer

Reply:
For `PermissionRequested` API, there are two factors. One is the presence of the `FrameCreated` handler, and the other is the need to handle the `PermissionRequested` API. Simply subscribing to the `FrameCreated` handler does not affect where the `PermissionRequested` event is actually handled.
Suppose there is a permission request from frame D. Developers want to handle the `PermissionRequested` event from the top-level iframe B in component I.  Meanwhile, in component II, developers have an "app usage telemetry" that hooks all frame activity. __Unless__ the developers subscribe to the permission request and [Handled](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2permissionrequestedeventargs.handled?view=webview2-dotnet-1.0.2903.40#microsoft-web-webview2-core-corewebview2permissionrequestedeventargs-handled) in either frame D or frame C,  otherwise the `PermissionRequested`  event will still bubble up to the top-level iframe and handle in component I.
```
// Example:
//    A (main frame/CoreWebView2)
//    |
//    B (first-level iframe/CoreWebView2Frame)
//    |  
//    C (nested iframe)
//    |
//    D (nested iframe)
```

But I think this kind of shared behavior should be avoid by developers. Even if we, align with the browser, track all iframes by default and the `PermissionRequested` event starts to fire on the iframe where it originates, there could be shared behavior issue. If developers handle the `PermissionRequested` event in a nested iframe within one component, then another component won't have the opportunity to handle the `PermissionRequested` event at the top-level iframe.

Reply:
No issue after discussion: event bubbles out in existing event bubbling pattern.